orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/servergroup/strategies/RedBlackStrategy.groovy

```diff
@@ -50,17 +52,32 @@ class RedBlackStrategy implements Strategy, ApplicationContextAware {
 
   @Override
   List<Stage> composeFlow(Stage stage) {
-    def stages = []
-    def stageData = stage.mapTo(StageData)
-    def cleanupConfig = AbstractDeployStrategyStage.CleanupConfig.fromStage(stage)
-
-    Map baseContext = [
-      (cleanupConfig.location.singularType()): cleanupConfig.location.value,
-      cluster                                : cleanupConfig.cluster,
-      credentials                            : cleanupConfig.account,
-      moniker                                : cleanupConfig.moniker,
-      cloudProvider                          : cleanupConfig.cloudProvider,
-    ]
+    List<Stage> stages = []
+    StageData stageData = stage.mapTo(StageData)
+    Map<String, Object> baseContext = AbstractDeployStrategyStage.CleanupConfig.toContext(stageData)
+
+    try {
+      Stage parentCreateServerGroupStage = stage.directAncestors().find() { it.type == CreateServerGroupStage.PIPELINE_CONFIG_TYPE }
+
+      if (parentCreateServerGroupStage.status == ExecutionStatus.NOT_STARTED) {
+        // No point in composing the flow if we are called to plan "beforeStages" since we don't have any STAGE_BEFOREs.
+        // Also, we rely on the the source server group task to have run already.
+        // In the near future we will move composeFlow into beforeStages and afterStages instead of the
+        // deprecated aroundStages
+        return []
+      }
+
+      StageData parentStageData = parentCreateServerGroupStage.mapTo(StageData)
+      StageData.Source sourceServerGroup = parentStageData.source
+
+      // Short-circuit if there is no source server group
+      if (sourceServerGroup == null || sourceServerGroup.serverGroupName == null) {
+        return []
+      }
+    } catch (Exception e) {
+      // This probably means there was no parent CreateServerGroup stage - which should never happen
+      throw new IllegalStateException("Failed to determine source server group from parent stage while planning red/black flow", e)
```

Comment:
oh, also... there is a currently known issue with this that you might want to check if it is still an issue in gce:
if you select `clone server group` in the UI this fails in AWS today

Reply:
Wow good call, I had no idea you could select a strategy when doing an ad-hoc clone from the UI (at least in GCE, we disable strategies when doing an ad-hoc create server group from the UI so I had assumed this was the same for cloning). Do you think your refactor will resolve this?

Reply:
Just pushed a commit that addresses this for red/black by checking for an ancestor stage that is either a deploy stage or a clone stage, which at least fixes this for GCE!

Reply:
i think we will still need something like this after refactor... to get the source. but seems like a reasonable fix for now (i will borrow it from you for AWS) :)